### PR TITLE
Refactor total stats calculation into shared fn

### PR DESCRIPTION
**Pull request by brooksprumo**
#### Problem

Calculating TotalAccountsStats is done in two places, and that code was duplicated, which we know is not ideal.

#### Summary of Changes

Refactor the calculation of TotalAccountsStats into a shared function. (I also added a few new stats too.)

### CODE REVIEW

**Review by brooksprumo — COMMENTED** [review]
Added some additional commentary on the code. Hopefully it's helpful!

File: ledger-tool/src/main.rs [hunk 1]
```diff
@@ -2509,26 +2509,22 @@ fn main() {
...
-                {
-                    total_accounts_stats.num_rent_exempt_accounts += 1;
-                }
+            let mut measure = Measure::start("calculating total accounts stats");
+            let total_accounts_stats = bank::calculate_total_accounts_stats(
+                &bank,
+                accounts
+                    .iter()
+                    .map(|(pubkey, (account, _slot))| (pubkey, account)),
+            );
```
[the remark this answers is not in the document]
Comment (CriesofCarrots) [reply ledger-tool/src/main.rs:2518]: I'm wondering if we can see some improvement by getting rid of that first `collect()`, though...
Would something like this work, and if so, how would the total time compare?
```
let print_account_contents = !arg_matches.is_present("no_account_contents");
let accounts_iterator = bank
    .get_all_accounts_with_modified_slots()
    .unwrap()
    .into_iter()
    .filter(|(pubkey, _account, _slot)| {
        include_sysvars || !solana_sdk::sysvar::is_sysvar_id(pubkey)
    })
    .map(|(pubkey, account, slot)| {
        if print_account_contents {
            ... all the print logic ...
        }
        (pubkey, account)
});
let total_accounts_stats = ...
```

File: runtime/src/bank.rs [hunk 2]
```diff
@@ -6322,31 +6322,49 @@ impl Bank {
     }
 }
 
-/// Scan all the accounts for this bank and collect stats
+/// Get all the accounts for this bank and calculate stats
 pub fn get_total_accounts_stats(bank: &Bank) -> ScanResult<TotalAccountsStats> {
+    // This implementation could have used scan_accounts() directly, but does not so that it could
+    // share `calculate_total_accounts_stats()` with ledger-tool (see subcommand "accounts").
```
Comment (CriesofCarrots) [inline runtime/src/bank.rs:6328]: I think scrap this comment, as it's clear from the implementation and callers.

File: runtime/src/bank.rs [hunk 3]
```diff
@@ -6322,31 +6322,49 @@ impl Bank {
     }
 }
 
-/// Scan all the accounts for this bank and collect stats
+/// Get all the accounts for this bank and calculate stats
 pub fn get_total_accounts_stats(bank: &Bank) -> ScanResult<TotalAccountsStats> {
```
[the remark this answers is not in the document]
Comment (brooksprumo) [reply runtime/src/bank.rs]: Yep, done.

File: runtime/src/bank.rs (hunk 2 above)
Comment (brooksprumo) [reply runtime/src/bank.rs]: Done.

**Comment by brooksprumo** [comment]
@CriesofCarrots Replying to: https://github.com/solana-labs/solana/pull/21212#discussion_r745951662

> I'm wondering if we can see some improvement by getting rid of that first collect(), though...
Would something like this work, and if so, how would the total time compare?

My first attempt was to do something close: first I'd get all the accounts and filter(), then call `calculate()` on (1), and finally do the map & collect then print account contents. This resulted in two calls to `get_all_accounts()`, which was a dealbreaker.

I did wonder if the BTree was _really_ necessary. I think your impl will work great, but the pubkeys won't be sorted when printing the account contents. Does that matter?

**Review by CriesofCarrots — APPROVED** [review]
> I did wonder if the BTree was really necessary. I think your impl will work great, but the pubkeys won't be sorted when printing the account contents. Does that matter?

Ah, I did forget about the sorting. I tend to think this sort is not super important; since there are so many accounts, it's not particularly helpful to be able to scan the output by pubkey anyway. It might actually be useful to have customizable sorts (by program id, or data size, etc) at some point in the future. But I'm okay going with this for now to preserve existing behavior.

**Pull request merged**